source/slang/ir.cpp

```diff
@@ -5251,6 +5381,47 @@ namespace Slang
         IRSpecEnv globalEnv;
     };
 
+    struct IRGenericSpecKey
+    {
+        // Note: Slang::Dictionary requires key types to have default constructors
+        IRGenericSpecKey()
+        {}
+
+        IRGenericSpecKey(IRSpecialize* specializeInst)
+        {
+            m_values.Add(specializeInst->getBase());
+            auto argCount = specializeInst->getArgCount();
+            for(UInt aa = 0; aa < argCount; ++aa)
+            {
+                m_values.Add(specializeInst->getArg(aa));
+            }
+        }
+
+        List<IRInst*> m_values;
+
+        bool operator==(IRGenericSpecKey const& other) const
+        {
+            auto valueCount = m_values.Count();
+            if(valueCount != other.m_values.Count()) return false;
+            for(UInt ii = 0; ii < valueCount; ++ii)
+            {
+                if(m_values[ii] != other.m_values[ii]) return false;
```

Comment:
The current status of the compiler is a bit off from ideal here, but the test in this case represents what I want it to be in the long run.

Rhe `IRBuilder` (through `SharedIRBuilder`) goes to some lengths to try to ensure that it doesn't create distinct `IRConstant` nodes that have the same type and value. For that matter, it *also* tries to make sure that it doesn't create two different `IRType` nodes that both represent the same time.

There are places in the code, however, where we might create new `IRConstant`s or types as part of some pass, and those would only end up being unique'd within the `SharedIRBuilder` used by the given pass, and might not be globally unique. When that happens, we could accidentally instantiate a generic more than once given logic like that above.

In the case of generic specialization this isn't expected to happen in practice because the only step that precedes generic specialization is the "linking" step that is creating fresh IR (using a single `SharedIRBuilder`) by copying from other modules. That step already needs to deduplicate things because different modules can obviously have different constants with the same value.

Still, in the long run we either need to make the deduplication more global (so that it belongs to the `IRModule` and we somehow deal with issues around remove/deletion of instructions that might also be cached in a dictionary), or introduce an explicit pass to remove duplicates which we invoke before any pass that really cares about deduplication.

What I do *not* want to do is introduce more complex "structural" equality rules for things like this. At least not if there is any way to avoid it...